pennylane/qinfo/fisher.py

```diff
@@ -0,0 +1,235 @@
+    we can compute a rescaled gradient using the CFIM. In this scenario, typically a Hamiltonian objective function :math:`\langle H \rangle` is minimized:
+
+    .. code-block:: python
+
+        H = qml.Hamiltonian(coeffs = [0.5, 0.5], ops = [qml.PauliZ(0), qml.PauliZ(1)])
+
+        @qml.qnode(dev)
+        def circ(params):
+            qml.RX(params[0], wires=0)
+            qml.RY(params[1], wires=0)
+            qml.RX(params[2], wires=1)
+            qml.RY(params[3], wires=1)
+            qml.CNOT(wires=(0,1))
+            return qml.expval(H)
+
+        params = pnp.random.random(4)
+
+    We can compute both the gradient of :math:`\langle H \rangle` and the CFIM with the same :class:`.QNode` ``circ`` in this example since ``classical_fisher()`` ignores the return types
+    and assumes ``qml.probs()`` for all wires.
+
+    >>> grad = qml.grad(circ)(params)
+    >>> cfim = qml.qinfo.classical_fisher(circ)(params)
+    >>> print(grad.shape, cfim.shape)
+    (4,) (4, 4)
+
+    Combined together, we can get a rescaled gradient to be employed for optimization schemes like natural gradient descent.
+
+    >>> rescaled_grad = cfim @ grad
+    >>> print(rescaled_grad)
+    [-0.66772533 -0.16618756 -0.05865127 -0.06696078]
+
+    """
+    new_qnode = _make_probs(qnode, post_processing_fn=lambda x: qml.math.squeeze(qml.math.stack(x)))
+
+    interface = qnode.interface
+
+    if interface in ("jax", "jax-jit"):
+        import jax
+
+        jac = jax.jacobian(new_qnode, argnums=argnums)
+
+    if interface == "torch":
+        jac = _torch_jac(new_qnode)
+
+    if interface == "autograd":
+        jac = qml.jacobian(new_qnode)
+
+    if interface == "tf":
+        jac = _tf_jac(new_qnode)
+
+    def wrapper(*args, **kwargs):
+        j = jac(*args, **kwargs)
+        p = new_qnode(*args, **kwargs)
+
+        # In case multiple variables are used
+        if isinstance(j, tuple) and len(j) > 1:
+            res = []
+            for j_i in j:
+                if interface == "tf":
+                    j_i = qml.math.transpose(qml.math.cast(j_i, dtype=p.dtype))
+
+                res.append(_compute_cfim(p, j_i))
+
+            return res
+
+        return _compute_cfim(p, j)
+
+    return wrapper
+
+
+def _compute_cfim(p, dp):
+    r"""Computes the (num_params, num_params) classical fisher information matrix from the probabilities and its derivatives
+    I.e. it computes :math:`classical_fisher_{ij} = \sum_\ell (\partial_i p_\ell) (\partial_i p_\ell) / p_\ell`
+    """
+    # Note that casting and being careful about dtypes is necessary as interfaces
+    # typically treat derivatives (dp) with float32, while standard execution (p) comes in float64
+
+    nonzeros_p = qml.math.where(p > 0, p, qml.math.ones_like(p))
+    one_over_p = qml.math.where(p > 0, qml.math.ones_like(p), qml.math.zeros_like(p))
+    one_over_p = qml.math.divide(one_over_p, nonzeros_p)
```

Comment:
Yes indeed! I just tried to write a brief explanation but it is actually not that straight forward. I would argue this is the way to do it because in the derivation, appendix A in [2103.15191](https://arxiv.org/pdf/2103.15191.pdf), it is implicitly assumed that the summation only goes over the non-zero values of $p_\ell$ (or, alternatively, that $0\log(0)=0$).

So here we would have something like $p_\ell(\theta) (\partial_i \partial_j \log(p_\ell(\theta))) |_{p=0} = 0$

Reply:
Ok perfect thanks!